packages/adaptive_scaffold/lib/src/adaptive_layout.dart

```diff
@@ -0,0 +1,517 @@
+// found in the LICENSE file.
+
+import 'dart:ui';
+import 'package:flutter/widgets.dart';
+import 'breakpoints.dart';
+import 'slot_layout.dart';
+
+enum _SlotIds {
+  primaryNavigation,
+  secondaryNavigation,
+  topNavigation,
+  bottomNavigation,
+  body,
+  secondaryBody,
+}
+
+/// [AdaptiveLayout] is dedicated to simplifying the process of developing
+/// layouts for many different screen sizes. It separates the screen into
+/// several sections, also referred to as "slots". The Widget defines slots for
+/// the different placements of navigational elements and two slots for a body
+/// and secondaryBody. All of these slots take a [SlotLayout] Widget. This class
+/// works in junction with [SlotLayout] and [SlotLayoutConfig]. This class
+/// handles the placement of the slots on the screen and animations regarding
+/// their macromovements.
+///
+/// See also:
+///
+///  * [SlotLayout], which handles the actual switching and animations between
+/// elements based on [Breakpoint]s.
+///  * [SlotLayoutConfig], which holds information regarding the actual Widgets
+/// and the desired way to animate between switches. Often used within
+/// [SlotLayout].
+///  * [Design Doc](https://flutter.dev/go/adaptive-layout-foldables).
+///  * [Material Design 3 Specifications] (https://m3.material.io/foundations/adaptive-design/overview).
+class AdaptiveLayout extends StatefulWidget {
+  /// Creates an [AdaptiveLayout] widget.
+  const AdaptiveLayout({
+    this.primaryNavigation,
+    this.secondaryNavigation,
+    this.topNavigation,
+    this.bottomNavigation,
+    this.body,
+    this.secondaryBody,
+    this.bodyRatio,
+    this.internalAnimations = true,
+    this.bodyOrientation = Axis.horizontal,
+    super.key,
+  });
+
+  /// The slot placed on the beginning side of the screen; meaning on the left
+  /// when the textDirection is LTR and on the right when it is RTL.
+  ///
+  /// Note: if using flexibly sized Widgets like [Container], wrap the Widget in a
+  /// [SizedBox] or limit its size by another method.
+  final SlotLayout? primaryNavigation;
+
+  /// The slot placed on the end side of the screen; meaning on the right
+  /// when the textDirection is LTR and on the left when it is RTL.
+  ///
+  /// Note: if using flexibly sized Widgets like [Container], wrap the Widget in a
+  /// [SizedBox] or limit its size by another method.
+  final SlotLayout? secondaryNavigation;
+
+  /// The slot placed on the top part of the screen.
+  ///
+  /// Note: if using flexibly sized Widgets like [Container], wrap the Widget in a
+  /// [SizedBox] or limit its size by another method.
+  final SlotLayout? topNavigation;
+
+  /// The slot placed on the bottom part of the screen.
+  ///
+  /// Note: if using flexibly sized Widgets like [Container], wrap the Widget in a
+  /// [SizedBox] or limit its size by another method.
+  final SlotLayout? bottomNavigation;
+
+  /// The slot that fills the rest of the space in the center.
+  final SlotLayout? body;
```

Comment:
When would an app have a null body?

Reply:
It never really would but everything still works as expected without a body so I wouldn't directly require it. Having the widget be as customizable as possible just prevents a developer from asking for us to change the API later to fit their odd usecase. If you feel like it would be better to strictly require a body though, let me know and I'll make that required. Thanks.

Reply:
Supporting a nullable body is OK. I thought that maybe you had a specific use-case in mind.